File: patterns/01-design-patterns/01-creational/01-factory-method/js/factory_method.mjs
export class PatternError extends Error {
  constructor(code) { super(code); this.code = code; }
}

class CsvFormatter {
  get mediaType() { return "text/csv"; }
  render(records) {
    return ["id,name", ...records.map((record) => `${record.id},${record.name}`)].join("\n");
  }
}

class JsonFormatter {
  get mediaType() { return "application/json"; }
  render(records) { return records.map((record) => ({ ...record })); }
}

class ExportJob {
  createFormatter() { throw new Error("abstract factory method"); }
  export(records) {
    const formatter = this.createFormatter();
    return { media_type: formatter.mediaType, body: formatter.render(records) };
  }
}

class CsvExportJob extends ExportJob { createFormatter() { return new CsvFormatter(); } }
class JsonExportJob extends ExportJob { createFormatter() { return new JsonFormatter(); } }

export function evaluate(input) {
  const creators = { csv: CsvExportJob, json: JsonExportJob };
  const Creator = creators[input.format];
  if (Creator === undefined) throw new PatternError("unsupported_format");
  return new Creator().export(input.records ?? []);
}

File: patterns/01-design-patterns/01-creational/01-factory-method/python/factory_method.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Protocol


class PatternError(Exception):
    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


class Formatter(Protocol):
    media_type: str
    def render(self, records: list[dict[str, str]]) -> Any: ...


class CsvFormatter:
    media_type = "text/csv"

    def render(self, records):
        rows = ["id,name"]
        rows.extend(f"{record['id']},{record['name']}" for record in records)
        return "\n".join(rows)


class JsonFormatter:
    media_type = "application/json"

    def render(self, records):
        return [dict(record) for record in records]


class ExportJob(ABC):
    @abstractmethod
    def create_formatter(self) -> Formatter: ...

    def export(self, records):
        formatter = self.create_formatter()
        return {"media_type": formatter.media_type, "body": formatter.render(records)}


class CsvExportJob(ExportJob):
    def create_formatter(self):
        return CsvFormatter()


class JsonExportJob(ExportJob):
    def create_formatter(self):
        return JsonFormatter()


def evaluate(input_data):
    creators = {"csv": CsvExportJob, "json": JsonExportJob}
    creator_type = creators.get(input_data.get("format"))
    if creator_type is None:
        raise PatternError("unsupported_format")
    return creator_type().export(input_data.get("records", []))
